runners/direct-java/src/test/java/org/apache/beam/runners/direct/WriteWithShardingFactoryTest.java

```diff
@@ -0,0 +1,230 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.beam.runners.direct;
+
+import static org.hamcrest.Matchers.anyOf;
+import static org.hamcrest.Matchers.containsInAnyOrder;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.greaterThan;
+import static org.hamcrest.Matchers.lessThan;
+import static org.hamcrest.Matchers.not;
+import static org.junit.Assert.assertThat;
+
+import org.apache.beam.runners.direct.WriteWithShardingFactory.KeyBasedOnCountFn;
+import org.apache.beam.sdk.coders.VarLongCoder;
+import org.apache.beam.sdk.io.Sink;
+import org.apache.beam.sdk.io.TextIO;
+import org.apache.beam.sdk.io.Write;
+import org.apache.beam.sdk.options.PipelineOptions;
+import org.apache.beam.sdk.testing.TestPipeline;
+import org.apache.beam.sdk.transforms.Create;
+import org.apache.beam.sdk.transforms.DoFnTester;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.util.IOChannelUtils;
+import org.apache.beam.sdk.util.PCollectionViews;
+import org.apache.beam.sdk.util.WindowingStrategy;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PCollectionView;
+import org.apache.beam.sdk.values.PDone;
+
+import com.google.common.base.Function;
+import com.google.common.collect.Iterables;
+
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+
+import java.io.File;
+import java.io.FileReader;
+import java.io.Reader;
+import java.nio.CharBuffer;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.UUID;
+import java.util.concurrent.ThreadLocalRandom;
+
+import javax.annotation.Nullable;
+
+/**
+ * Created by tgroh on 7/14/16.
+ */
+public class WriteWithShardingFactoryTest {
+  @Rule public TemporaryFolder tmp = new TemporaryFolder();
+  private WriteWithShardingFactory factory = new WriteWithShardingFactory();
+
+  @Test
+  public void dynamicallyReshardedWrite() throws Exception {
+    List<String> strs = new ArrayList<>(10000);
+    for (int i = 0; i < 10000; i++) {
+      strs.add(UUID.randomUUID().toString());
+    }
+    Collections.shuffle(strs);
+
+    String fileName = "resharded_write";
+    String outputPath = tmp.getRoot().getAbsolutePath();
+    String targetLocation = IOChannelUtils.resolve(outputPath, fileName);
+    TestPipeline p = TestPipeline.create();
+    // TextIO is implemented in terms of the Write PTransform. When sharding is not specified,
+    // resharding should be automatically applied
+    p.apply(Create.of(strs)).apply(TextIO.Write.to(targetLocation));
+
+    p.run();
+
+    Collection<String> files = IOChannelUtils.getFactory(outputPath).match(targetLocation + "*");
+    assertThat(files.size(), greaterThan(1));
+    assertThat(files.size(), lessThan((int) (Math.log10(10000) + 1)));
```

Comment:
Indeed.

In fact, this can all be nested into a single assertion. Did that as well.